Index: command.c
===================================================================
RCS file: /home/ncvs/src/usr.sbin/ppp/command.c,v
retrieving revision 1.103
diff -c -r1.103 command.c
*** command.c	1997/11/18 00:19:28	1.103
--- command.c	1997/11/18 19:35:36
***************
*** 1013,1019 ****
    void (*DiscardAll)(void);
  
    res = 0;
!   if (strcasecmp(argv[0], "local")) {
      Discard = LogDiscard;
      Keep = LogKeep;
      DiscardAll = LogDiscardAll;
--- 1013,1019 ----
    void (*DiscardAll)(void);
  
    res = 0;
!   if (argc == 0 || strcasecmp(argv[0], "local")) {
      Discard = LogDiscard;
      Keep = LogKeep;
      DiscardAll = LogDiscardAll;
