Index: libf2c/libI77/lread.c
===================================================================
RCS file: /cvs/egcs/egcs/libf2c/libI77/lread.c,v
retrieving revision 1.4
diff -u -p -r1.4 lread.c
--- lread.c	1999/03/17 08:21:19	1.4
+++ lread.c	1999/07/11 03:42:33
@@ -70,7 +70,18 @@ un_getc(int x, FILE *f__cf)
 #endif
 { return ungetc(x,f__cf); }
 #else
-#define un_getc ungetc
+ /* The OpenServer 5.0.5 dynamic linker aborts on g77 programs 
+  * that link against this object.   Simply changing the shape of 
+  * this relocation is enough to circumvent the bug in the dynamic 
+  * linker.
+  */
+#if 0
+ #define un_getc ungetc
+#else
+ typedef int (blah) (int c, FILE *stream);
+ blah *un_getc = ungetc;
+#endif
+
 #ifdef KR_headers
  extern int ungetc();
 #else
