Index: gcc/ChangeLog
from  Alexandre Oliva  <oliva@lsd.ic.unicamp.br>
	
	* config/i386/sol2gas.h: I had installed the wrong version of Sept
	22's patch in the release branch.  Fixed.
	* configure.in: Likewise.
	* configure: Rebuilt.
	Reported by Jason Beardsley <jbeardsley@origin.ea.com>
	
Index: gcc/configure.in
===================================================================
RCS file: /cvs/gcc/egcs/gcc/configure.in,v
retrieving revision 1.246.4.12
diff -u -r1.246.4.12 configure.in
--- gcc/configure.in	1999/10/13 07:58:02	1.246.4.12
+++ gcc/configure.in	1999/10/28 23:05:38
@@ -1399,11 +1399,10 @@
 changequote([,])dnl
 		xm_file="xm-siglist.h xm-alloca.h ${xm_file}"
 		xm_defines="USG POSIX SMALL_ARG_MAX"
+		tm_file=i386/sol2.h
 		if test x$gas = xyes; then
 			# Only needed if gas does not support -s
-			tm_file=i386/sol2gas.h
-		else
-			tm_file=i386/sol2.h
+			tm_file="i386/sol2gas.h ${tm_file}"
 		fi
 		tmake_file=i386/t-sol2
 		extra_parts="crt1.o crti.o crtn.o gcrt1.o gmon.o crtbegin.o crtend.o"
Index: gcc/config/i386/sol2gas.h
===================================================================
RCS file: /cvs/gcc/egcs/gcc/config/i386/sol2gas.h,v
retrieving revision 1.1.2.1
diff -u -r1.1.2.1 sol2gas.h
--- gcc/config/i386/sol2gas.h	1999/09/22 10:44:21	1.1.2.1
+++ gcc/config/i386/sol2gas.h	1999/10/28 23:05:38
@@ -4,8 +4,6 @@
    Copyright (C) 1999 Free Software Foundation, Inc.
 */
 
-#ifndef GAS_DOES_NOT_SUPPORT_MINUS_S
-#define GAS_DOES_NOT_SUPPORT_MINUS_S 1
+#ifndef GAS_REJECTS_MINUS_S
+#define GAS_REJECTS_MINUS_S 1
 #endif
-
-#include "i386/sol2.h"
Index: gcc/configure
===================================================================
RCS file: /cvs/gcc/egcs/gcc/configure,v
retrieving revision 1.242.4.12
diff -u -r1.242.4.12 configure
--- gcc/configure	1999/10/13 07:58:02	1.242.4.12
+++ gcc/configure	1999/10/28 23:10:32
@@ -3750,11 +3750,11 @@
 	i[34567]86-*-solaris2*)
 		xm_file="xm-siglist.h xm-alloca.h ${xm_file}"
 		xm_defines="USG POSIX SMALL_ARG_MAX"
+		tm_file=i386/sol2.h
 		if test x$gas = xyes; then
 			# Only needed if gas does not support -s
-			tm_file=i386/sol2gas.h
-		else
-			tm_file=i386/sol2.h
+			tm_file="i386/sol2gas.h ${tm_file}"
+			#
 		fi
 		tmake_file=i386/t-sol2
 		extra_parts="crt1.o crti.o crtn.o gcrt1.o gmon.o crtbegin.o crtend.o"
