*** gcc-2.6.3/Makefile.in	Fri Dec  2 16:03:09 1994
--- g77-2.6.3-0.5.16/Makefile.in	Thu Aug 10 04:01:02 1995
*************** ENQUIRE_CFLAGS = -DNO_MEM -DNO_LONG_DOUB
*** 118,121 ****
--- 118,126 ----
  ENQUIRE_LDFLAGS = $(LDFLAGS)
  
+ # NEXT FOUR LINES ADDED BY g77 PATCH ONLY TO ENABLE COMPATIBILITY WITH 2.7.x.
+ # Sed command to transform gcc to installed name.  Overwritten by configure.
+ program_transform_name = -e s,x,x,
+ program_transform_cross_name = -e s,^,$(target)-,
+ 
  # Tools to use when building a cross-compiler.
  # These are used because `configure' appends `cross-make'
*************** infodir = $(prefix)/info
*** 176,179 ****
--- 181,188 ----
  # Extension (if any) to put in installed man-page filename.
  manext = .1
+ # NEXT THREE LINES ADDED BY g77 PATCH ONLY TO ENABLE COMPATIBILITY WITH 2.7.x.
+ objext = .o
+ exeext =
+ 
  # Directory in which to put man pages.
  mandir = $(prefix)/man/man1
*************** c-common.o : c-common.c $(CONFIG_H) $(TR
*** 1010,1014 ****
  # Language-independent files.
  
! gcc.o: gcc.c $(CONFIG_H) multilib.h config.status
  	$(CC) $(ALL_CFLAGS) $(ALL_CPPFLAGS) $(INCLUDES) \
    -DSTANDARD_STARTFILE_PREFIX=\"$(libdir)/\" \
--- 1019,1023 ----
  # Language-independent files.
  
! gcc.o: gcc.c $(CONFIG_H) multilib.h config.status f/lang-specs.h
  	$(CC) $(ALL_CFLAGS) $(ALL_CPPFLAGS) $(INCLUDES) \
    -DSTANDARD_STARTFILE_PREFIX=\"$(libdir)/\" \
*************** stor-layout.o : stor-layout.c $(CONFIG_H
*** 1032,1036 ****
  fold-const.o : fold-const.c $(CONFIG_H) $(TREE_H) flags.h 
  toplev.o : toplev.c $(CONFIG_H) $(TREE_H) $(RTL_H) flags.h input.h \
!    insn-attr.h xcoffout.h defaults.h
  	$(CC) $(ALL_CFLAGS) $(ALL_CPPFLAGS) $(INCLUDES) \
  	  $(MAYBE_TARGET_DEFAULT) $(MAYBE_USE_COLLECT2) \
--- 1041,1045 ----
  fold-const.o : fold-const.c $(CONFIG_H) $(TREE_H) flags.h 
  toplev.o : toplev.c $(CONFIG_H) $(TREE_H) $(RTL_H) flags.h input.h \
!    insn-attr.h xcoffout.h defaults.h f/lang-options.h
  	$(CC) $(ALL_CFLAGS) $(ALL_CPPFLAGS) $(INCLUDES) \
  	  $(MAYBE_TARGET_DEFAULT) $(MAYBE_USE_COLLECT2) \
*** gcc-2.6.3/fold-const.c	Fri Dec  2 16:03:43 1994
--- g77-2.6.3-0.5.16/fold-const.c	Thu Aug 10 04:01:03 1995
*************** eval_subst (arg, old0, new0, old1, new1)
*** 1938,1943 ****
--- 1938,1948 ----
        switch (code)
  	{
+   /* ??? Disable this since the SAVE_EXPR might already be in use outside
+      the expression.  There may be no way to make this work, but it needs
+      to be looked at again for 2.6.  */
+ #if 0
  	case SAVE_EXPR:
  	  return eval_subst (TREE_OPERAND (arg, 0), old0, new0, old1, new1);
+ #endif
  
  	case COMPOUND_EXPR:
*************** fold (expr) 
*** 3953,3958 ****
--- 3958,3968 ----
  	  tree xarg0 = arg0;
  
+   /* ??? Disable this since the SAVE_EXPR might already be in use outside
+      the expression.  There may be no way to make this work, but it needs
+      to be looked at again for 2.6.  */
+ #if 0
  	  if (TREE_CODE (xarg0) == SAVE_EXPR)
  	    have_save_expr = 1, xarg0 = TREE_OPERAND (xarg0, 0);
+ #endif
  
  	  STRIP_NOPS (xarg0);
*************** fold (expr) 
*** 3971,3976 ****
--- 3981,3991 ----
  	    }
  
+   /* ??? Disable this since the SAVE_EXPR might already be in use outside
+      the expression.  There may be no way to make this work, but it needs
+      to be looked at again for 2.6.  */
+ #if 0
  	  if (TREE_CODE (xarg0) == SAVE_EXPR)
  	    have_save_expr = 1, xarg0 = TREE_OPERAND (xarg0, 0);
+ #endif
  
  	  STRIP_NOPS (xarg0);
*** gcc-2.6.3/gcc.c	Mon Nov  7 11:01:43 1994
--- g77-2.6.3-0.5.16/gcc.c	Thu Aug 10 04:01:03 1995
*************** static struct compiler default_compilers
*** 712,715 ****
--- 712,716 ----
  					      %{c:%W{o*}%{!o*:-o %w%b.o}}\
  					      %{!c:-o %d%w%u.o} %{!pipe:%g.s} %A\n}}}}}} "},
+ #include "f/lang-specs.h"
    /* Mark end of table */
    {0, 0}
*** gcc-2.6.3/obstack.c	Sat Nov 12 06:40:02 1994
--- g77-2.6.3-0.5.16/obstack.c	Thu Aug 10 04:01:04 1995
*************** Foundation, 675 Mass Ave, Cambridge, MA 
*** 27,33 ****
     (especially if it is a shared library).  Rather than having every GNU
     program understand `configure --with-gnu-libc' and omit the object files,
!    it is simpler to just do this in the source for each such file.  */
  
! #if defined (_LIBC) || !defined (__GNU_LIBRARY__)
  
  
--- 27,40 ----
     (especially if it is a shared library).  Rather than having every GNU
     program understand `configure --with-gnu-libc' and omit the object files,
!    it is simpler to just do this in the source for each such file.
  
!    Actually, don't comment this code out after all.  Else, unless the
!    inlining set up by obstack.h also is commented out (and replaced by
!    including the system's <obstack.h>), bugs will result on a system
!    with an older obstack (with older and different inlining) installed.
!    And for now it doesn't seem worth having obstack.h #include <stdio.h>,
!    as above, just to get __GNU_LIBRARY__ defined, for example.  */
! 
! #if defined (_LIBC) || !defined (__GNU_LIBRARY__) || 1
  
  
*** gcc-2.6.3/stor-layout.c	Thu Oct 20 09:41:02 1994
--- g77-2.6.3-0.5.16/stor-layout.c	Thu Aug 10 04:01:04 1995
*************** get_pending_sizes ()
*** 80,83 ****
--- 80,93 ----
  }
  
+ void
+ put_pending_sizes (chain)
+      tree chain;
+ {
+   if (pending_sizes)
+     abort ();
+ 
+   pending_sizes = chain;
+ }
+ 
  /* Given a size SIZE that isn't constant, return a SAVE_EXPR
     to serve as the actual size-expression for a type or decl.  */
*** gcc-2.6.3/toplev.c	Tue Oct 25 16:09:12 1994
--- g77-2.6.3-0.5.16/toplev.c	Thu Aug 10 04:01:05 1995
*************** char *lang_options[] =
*** 725,728 ****
--- 725,730 ----
    "-Wno-protocol",
  
+ #include "f/lang-options.h"
+ 
    /* This is for GNAT and is temporary.  */
    "-gnat",
*** gcc-2.6.3/tree.c	Fri Dec  2 16:03:49 1994
--- g77-2.6.3-0.5.16/tree.c	Thu Aug 10 04:01:05 1995
*************** save_expr (expr)
*** 1984,1988 ****
  
    if (TREE_CONSTANT (t) || (TREE_READONLY (t) && ! TREE_SIDE_EFFECTS (t))
!       || TREE_CODE (t) == SAVE_EXPR)
      return t;
  
--- 1984,1988 ----
  
    if (TREE_CONSTANT (t) || (TREE_READONLY (t) && ! TREE_SIDE_EFFECTS (t))
!       || TREE_CODE (t) == SAVE_EXPR || TREE_CODE (t) == ERROR_MARK)
      return t;
  
*** gcc-2.6.3/tree.h	Thu Sep  8 14:25:41 1994
--- g77-2.6.3-0.5.16/tree.h	Thu Aug 10 04:01:06 1995
*************** extern tree size_int			PROTO((unsigned))
*** 1245,1248 ****
--- 1245,1249 ----
  extern tree round_up			PROTO((tree, int));
  extern tree get_pending_sizes		PROTO((void));
+ extern void put_pending_sizes		PROTO((tree));
  
  /* Type for sizes of data-type.  */
