*** gcc-2.7.0/obstack.c	Thu Jun 15 18:12:08 1995
--- g77-2.7.0-0.5.16/obstack.c	Thu Aug 10 04:06:51 1995
*************** Foundation, 675 Mass Ave, Cambridge, MA 
*** 27,33 ****
     (especially if it is a shared library).  Rather than having every GNU
     program understand `configure --with-gnu-libc' and omit the object files,
!    it is simpler to just do this in the source for each such file.  */
  
! #if defined (_LIBC) || !defined (__GNU_LIBRARY__)
  
  
--- 27,40 ----
     (especially if it is a shared library).  Rather than having every GNU
     program understand `configure --with-gnu-libc' and omit the object files,
!    it is simpler to just do this in the source for each such file.
  
!    Actually, don't comment this code out after all.  Else, unless the
!    inlining set up by obstack.h also is commented out (and replaced by
!    including the system's <obstack.h>), bugs will result on a system
!    with an older obstack (with older and different inlining) installed.
!    And for now it doesn't seem worth having obstack.h #include <stdio.h>,
!    as above, just to get __GNU_LIBRARY__ defined, for example.  */
! 
! #if defined (_LIBC) || !defined (__GNU_LIBRARY__) || 1
  
  
*** gcc-2.7.0/stor-layout.c	Thu Jun 15 08:08:47 1995
--- g77-2.7.0-0.5.16/stor-layout.c	Thu Aug 10 14:00:15 1995
*************** get_pending_sizes ()
*** 82,85 ****
--- 82,95 ----
  }
  
+ void
+ put_pending_sizes (chain)
+      tree chain;
+ {
+   if (pending_sizes)
+     abort ();
+ 
+   pending_sizes = chain;
+ }
+ 
  /* Given a size SIZE that may not be a constant, return a SAVE_EXPR
     to serve as the actual size-expression for a type or decl.  */
*** gcc-2.7.0/tree.c	Thu Jun 15 08:10:23 1995
--- g77-2.7.0-0.5.16/tree.c	Mon Aug 28 12:21:15 1995
*************** save_expr (expr)
*** 2111,2115 ****
  
    if (TREE_CONSTANT (t) || (TREE_READONLY (t) && ! TREE_SIDE_EFFECTS (t))
!       || TREE_CODE (t) == SAVE_EXPR)
      return t;
  
--- 2111,2115 ----
  
    if (TREE_CONSTANT (t) || (TREE_READONLY (t) && ! TREE_SIDE_EFFECTS (t))
!       || TREE_CODE (t) == SAVE_EXPR || TREE_CODE (t) == ERROR_MARK)
      return t;
  
*** gcc-2.7.0/tree.h	Thu Jun 15 08:10:49 1995
--- g77-2.7.0-0.5.16/tree.h	Thu Aug 10 04:06:52 1995
*************** extern tree size_int			PROTO((unsigned H
*** 1290,1293 ****
--- 1290,1294 ----
  extern tree round_up			PROTO((tree, int));
  extern tree get_pending_sizes		PROTO((void));
+ extern void put_pending_sizes		PROTO((tree));
  
  /* Type for sizes of data-type.  */
